include/aws/common/math.h

```diff
@@ -0,0 +1,189 @@
+#ifndef AWS_COMMON_math_H
+#define AWS_COMMON_math_H
+
+#include <stdint.h>
+#include <stdlib.h>
+
+/*
+ * Copyright 2010-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+/**
+ * Multiplies a * b. If the result overflows, returns 2^64 - 1.
+ */
+static inline uint64_t aws_common_mul_u64_saturating(uint64_t a, uint64_t b) {
+#if defined(__x86_64__) && defined(__GNUC__)
```

Comment:
As much as I love this, you need to convert this to msvc intrinsics for Windows builds.

Reply:
I'm not set up to test those currently.